modules/gapi/test/streaming/gapi_streaming_vpl_device_selector.cpp

```diff
@@ -0,0 +1,227 @@
+#ifdef HAVE_DIRECTX
+#ifdef HAVE_D3D11
+        std::vector<CfgParam> cfg_params_w_dx11;
+        cfg_params_w_dx11.push_back(CfgParam::create<uint32_t>("mfxImplDescription.AccelerationMode",
+                                                                MFX_ACCEL_MODE_VIA_D3D11));
+        Device::Ptr empty_device_ptr = nullptr;
+        Context::Ptr empty_ctx_ptr = nullptr;
+        EXPECT_THROW(CfgParamDeviceSelector sel(empty_device_ptr, empty_ctx_ptr,
+                                                cfg_params_w_dx11),
+                     std::logic_error); // empty_device_ptr must be valid
+#endif // HAVE_D3D11
+#endif // HAVE_DIRECTX
+    }
+
+    {
+#ifdef HAVE_DIRECTX
+#ifdef HAVE_D3D11
+        ID3D11Device *device = nullptr;
+        ID3D11DeviceContext* device_context = nullptr;
+        {
+            //Create device
+            UINT creationFlags = 0;//D3D11_CREATE_DEVICE_BGRA_SUPPORT;
+
+            D3D_FEATURE_LEVEL featureLevels[] = { D3D_FEATURE_LEVEL_11_1,
+                                                  D3D_FEATURE_LEVEL_11_0,
+                                                };
+            D3D_FEATURE_LEVEL featureLevel;
+
+            // Create the Direct3D 11 API device object and a corresponding context.
+            HRESULT err = D3D11CreateDevice(
+                    nullptr, // Specify nullptr to use the default adapter.
+                    D3D_DRIVER_TYPE_HARDWARE,
+                    nullptr,
+                    creationFlags, // Set set debug and Direct2D compatibility flags.
+                    featureLevels, // List of feature levels this app can support.
+                    ARRAYSIZE(featureLevels),
+                    D3D11_SDK_VERSION, // Always set this to D3D11_SDK_VERSION.
+                    &device, // Returns the Direct3D device created.
+                    &featureLevel, // Returns feature level of device created.
+                    &device_context // Returns the device immediate context.
+                    );
+            EXPECT_FALSE(FAILED(err));
+        }
+        std::unique_ptr<CfgParamDeviceSelector> selector_ptr;
+        std::vector<CfgParam> cfg_params_w_dx11;
+        cfg_params_w_dx11.push_back(CfgParam::create<uint32_t>("mfxImplDescription.AccelerationMode",
+                                                                MFX_ACCEL_MODE_VIA_D3D11));
+        EXPECT_NO_THROW(selector_ptr.reset(new CfgParamDeviceSelector(device, device_context,
+                                                                      cfg_params_w_dx11)));
+        IDeviceSelector::DeviceScoreTable devs = selector_ptr->select_devices();
+
+        EXPECT_EQ(devs.size(), 1);
+        test_eq(*devs.begin(), IDeviceSelector::Score::Max,
+                AccelType::DX11, device);
+
+        Context ctx = selector_ptr->select_context(devs);
+        EXPECT_EQ(ctx.get_ptr(), device_context);
+#endif // HAVE_D3D11
+#endif // HAVE_DIRECTX
+    }
+}
+}
+} // namespace opencv_test
+#endif // HAVE_ONEVPL
```

Comment:
Should the full test bodies be enclosed under `HAVE_` macros?

Reply:
I mean, once and for all

Reply:
..or the idea is to extend this with other flags as they arrive?

I'd prefer different tests cover different cases otherwise you can't say for sure what actually passed/failed in the test looking at its name when it has that lots of compile conditionals inside

Reply:
Do you mean
`TEST(OneVPL_Source_Device_Selector_CfgParam, PtrDeviceFromCfgParam)` 
then
`TEST(OneVPL_Source_Device_Selector_CfgParam, DX11_PtrDeviceFromCfgParam)`
then
`TEST(OneVPL_Source_Device_Selector_CfgParam, YYY_PtrDeviceFromCfgParam)`


agree, it makes sense